common/fixation_entry.go

```diff
@@ -0,0 +1,280 @@
+	if err != nil {
+		return utils.LavaError(ctx, ctx.Logger(), "SetEntry_cant_find_entry", map[string]string{"fs.prefix": fs.prefix, "index": index, "block": strconv.FormatUint(block, 10)}, "can't set non-existent entry")
+	}
+
+	// update the entry's data
+	entry.Data = marshaledEntryData
+
+	// marshal the entry
+	marshaledEntry := fs.cdc.MustMarshal(entry)
+
+	// set the entry
+	store.Set(byteKey, marshaledEntry)
+
+	return nil
+}
+
+// handleRefAction handles ref actions: increase refs, decrease refs or do nothing
+func (fs *FixationStore) handleRefAction(ctx sdk.Context, entry *types.Entry, refAction types.ReferenceAction) error {
+	// check if entry is nil
+	if entry == nil {
+		return utils.LavaError(ctx, ctx.Logger(), "handleRefAction_nil_entry", map[string]string{}, "can't handle reference action, entry is nil")
+	}
+
+	// handle the ref action
+	switch refAction {
+	case types.ADD_REFERENCE:
+		entry.Refcount += 1
+	case types.SUB_REFERENCE:
+		if entry.GetRefcount() > 0 {
+			entry.Refcount -= 1
+		} else {
+			return utils.LavaError(ctx, ctx.Logger(), "handleRefAction_sub_ref_from_non_positive_count", map[string]string{"refCount": strconv.FormatUint(entry.GetRefcount(), 10)}, "refCount is not larger than zero. Can't subtract refcount")
+		}
+	case types.DO_NOTHING:
+	}
+
+	return nil
+}
+
+// GetStoreKey returns the Fixation store's store key
+func (fs *FixationStore) GetStoreKey() sdk.StoreKey {
+	return fs.storeKey
+}
+
+// GetCdc returns the Fixation store's codec
+func (fs *FixationStore) GetCdc() codec.BinaryCodec {
+	return fs.cdc
+}
+
+// Getprefix returns the Fixation store's fixation key
+func (fs *FixationStore) GetPrefix() string {
+	return fs.prefix
+}
+
+// getUnmarshaledEntryForBlock gets an entry by block. Block doesn't have to be precise, it gets the closest entry version
+func (fs *FixationStore) getUnmarshaledEntryForBlock(ctx sdk.Context, index string, block uint64, refAction types.ReferenceAction) (*types.Entry, error) {
+	// get the relevant store using index
+	store := prefix.NewStore(ctx.KVStore(fs.storeKey), types.KeyPrefix(fs.createStoreKey(index)))
+
+	// init a reverse iterator
+	iterator := sdk.KVStoreReversePrefixIterator(store, []byte{})
+	defer iterator.Close()
+
+	// iterate over entries
+	for ; iterator.Valid(); iterator.Next() {
+		// unmarshal the entry
+		var entry types.Entry
+		fs.cdc.MustUnmarshal(iterator.Value(), &entry)
+
+		// if the entry's block is smaller than or equal to the requested block, unmarshal the entry's data
+		// since the user might not know the precise block that the entry was created on, we get the closest one
+		// we get from the past since this was the entry that was valid in the requested block (assume there's an
+		// entry in block 100 and block 200 and the user asks for the version of block 199. Since the block 200's
+		// didn't exist yet, we get the entry from block 100)
+		if entry.GetBlock() <= block {
+			err := fs.handleRefAction(ctx, &entry, refAction)
```

Comment:
Traditionally, reference counting can be subtle, hence it is common practice to be explicit in the code - so it is clear if you increase or decrease the count. It makes the intention - and the code - clearer.  

Specifically here are 3 uses: 
- Get and increase, 
- Get without increase (safe, because we know we hold reference already), 
- Put back and decrease. 

It's clearer to have explicit APIs accordingly, e.g. "GetEntry", "GetEntrySafe", and "PutEntry". They all should use the underlying common `getUnmarshaledEntryForBlock()`.

However, this function (`getUnmarshaledEntryForBlock()`) is a common internal helper and **should  not** by itself handle the refcount - that should be up to the caller. The caller would only e.g. increase the refcount after additional checks (otherwise, if an error occurs the caller would need to undo the reference count). See the comment in GetEntry().

Reply:
done